### PR TITLE
[Merged by Bors] - feat: port coherence tactic

### PR DESCRIPTION
**Pull request by kim-em**
This is a by-hand port of `category_theory.monoidal.coherence`, which is blocking things like https://tqft.net/mathlib4/2023-05-26/representation_theory.group_cohomology.basic.pdf.

It does not include a port of the coherence tactic for bicategories, but this is not actually used in mathlib yet. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Monoidal/Coherence.lean [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+  Term.elabTerm (← `(CategoryTheory.FreeMonoidalCategory.projectMap _root_.id _ _
+    (CategoryTheory.MonoidalCategory.LiftHom.lift $(← Term.exprToSyntax e)))) none
+
+/-- Coherence tactic for monoidal categories. -/
+def monoidal_coherence (g : MVarId) : TermElabM Unit := do
+  withOptions (fun opts => opts.setNat `synthInstance.maxSize
+    (max 256 (opts.getNat `synthInstance.maxSize))) do
+  -- TODO: is this `dsimp only` step necessary? It doesn't appear to be in the tests below.
+  let (ty, _) ← dsimp (← g.getType) (← Simp.Context.ofNames [] true)
+  let some (_, lhs, rhs) := ty.eq? | throwError "Not an equation of morphisms."
```
Comment (mcdoll) [inline Mathlib/CategoryTheory/Monoidal/Coherence.lean:208]: I think the canonical way of throwing an error is `throwTacticEx `monoidal_coherence g m!"foobar"`

File: Mathlib/CategoryTheory/Monoidal/Coherence.lean [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+
+/-- If either the lhs or rhs is not a composition, compose it on the right with an identity. -/
+def insertTrailingIds : TacticM Unit := do
+  let some (_, lhs, rhs) := (← getMainTarget).eq? | throwError "Not an equality."
+  match lhs.getAppFnArgs with
+  | (``CategoryStruct.comp, _) => pure ()
+  | _ => evalTactic (← `(tactic| apply insert_id_lhs))
+  match rhs.getAppFnArgs with
+  | (``CategoryStruct.comp, _) => pure ()
+  | _ => evalTactic (← `(tactic| apply insert_id_rhs))
```
Comment (mcdoll) [inline Mathlib/CategoryTheory/Monoidal/Coherence.lean:309]: I think this should be `def insertTrailingIds (mvarId : MVarId) : MetaM (List MVarId)` with the mvarId version of `apply` and then use `liftMetaTactic`

File: Mathlib/CategoryTheory/Monoidal/Coherence.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/CategoryTheory/Monoidal/Coherence.lean]: Thanks, I've adopted this.

File: Mathlib/CategoryTheory/Monoidal/Coherence.lean (hunk 2 above)
Comment (semorrison) [reply Mathlib/CategoryTheory/Monoidal/Coherence.lean]: Sure!

**Review by kmill — APPROVED** [review]
@semorrison I'm happy enough with this to merge it. Now that I've edited it last, I think that means you get to review it next (once it passes CI).

**Comment by kmill** [comment]
I just looked through it again and I'll go ahead and merge it -- it's going to be reviewed again anyway once the bicategory parts are PR'd, and I figure you'd prefer have this in sooner over having a chance to review my changes.

bors r+

**Pull request closed without merging**